### ISSUE TITLE
new rule - fstring syntax in non-fstring

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#8151]
```py
asdf = 1
"value is {asdf}" # the user probably meant to make this an f string
```

**Comment by dhruvmanila** [comment]
Interesting! Although not sure how to reliably detect this. One way could be to pass the string as a f-string to the parser and check if there are any `ExprFormattedValue` node in it.

**Comment by charliermarsh** [comment]
I'm worried about false positives on this one.

**Comment by DetachHead** [comment]
i feel like they would be rare as long as it only complains when the expression inside the `{}` refers to an existing variable

**Comment by MichaReiser** [comment]
I would find this useful. Coming from Rust and JavaScript, I wrote `"{value}" too many times to then notice my debug print statements are completely useless. So I went back and added the `f` in front of my strings.

**Comment by dhruvmanila** [comment]
> I'm worried about false positives on this one.

Can you give an example where there would be false positives? Maybe we could only highlight when the node within `{...}` is an identifier and skip everything else.

**Comment by zanieb** [comment]
My common use false positive

```python
x = "{foo}"
x.format(foo="bar")
```

**Comment by DetachHead** [comment]
@zanieb in that case it shouldn't trigger a false positive if the rule works how i suggested in https://github.com/astral-sh/ruff/issues/8151#issuecomment-1786278064, because there's no variable in scope called `foo`.

imo the `format` method should only be used instead of f-strings in cases where the variables being formatted are not in scope. i've updated the OP to clarify this

**Comment by zanieb** [comment]
I agree requiring the variables to be in scope is a reasonable heuristic for avoiding false positives. I'm not sure I can come up with a common one that meets that requirement.

I found https://github.com/PrefectHQ/prefect/blob/f654a1b23eb5662afb5b8bfd08c9c86668e817ee/tests/agent/test_agent_queue_selection.py#L36 but that probably ought to be an f-string anyway. We can probably hint something else when `format` is being called instead of using a `f-string`.

**Comment by dhruvmanila** [comment]
We could just skip if the said string is part of a large `.format` call or `%` operator.

I think then we could support this as:
1. Skip if this string contains `{` / `}`
2. Skip if it's part of `.format` call or `%` operator
3. Parse the string with the `f` prefix
4. Visit each `ExprFormattedValue` node. If there are none, return
5. Create a diagnostic message only if _all_ of the expression inside the formatted node is
	a. An identifier
	b. It's defined in the scope

What do you think? @zanieb

**Comment by peterjc** [comment]
I'd welcome this. Cross reference https://github.com/peterjc/flake8-sfs/issues/5

See also the opposite request, f-strings without any variables in them - briefly implemented as flake8-pie's PIE782 code but withdrawn due to false positives. Cross reference https://github.com/peterjc/flake8-sfs/issues/3 and https://github.com/sbdchd/flake8-pie/issues/24

**Comment by dhruvmanila** [comment]
> See also the opposite request, f-strings without any variables in them

We do support this - https://docs.astral.sh/ruff/rules/f-string-missing-placeholders/ :)

**Comment by hauntsaninja** [comment]
pyanalyze supports this check: https://github.com/quora/pyanalyze/blob/647c64e8c9b4eaa81a5309d8e76e72bafe61d687/pyanalyze/name_check_visitor.py#L3096

Here are the heuristics it uses to avoid false positives:
- At least one name is used in the expression and all names used inside the expression are bound
- The string literal is not immediately `.format()`-ed
- It's not standalone expression (i.e. doesn't look like a docstring)
- It's not in a call that passes kwargs corresponding to all names (to avoid false positives with interpolation functions that work like `.format`)

**Comment by charliermarsh** [comment]
I'm supportive of adding this with the above heuristics.
